hash_table/hash_table.go

```diff
@@ -0,0 +1,134 @@
+package hash_table
+
+import (
+	"errors"
+)
+
+type Node struct {
+	key   int
+	value int
+	next  *Node
+}
+
+type HashTable struct {
+	arr  []*Node
+	size int
+}
+
+func NewHashTable() *HashTable {
+	return &HashTable{
+		arr: make([]*Node, 10),
+	}
+}
+
+func (h *HashTable) Insert(key int, value int) {
+	// same load factor as Java 10
+	if h.getLoadFactor() >= 0.75 {
+		h.rehash(true)
+	}
+
+	h.size++
+	hash := h.getHash(key)
+	node := h.arr[hash]
+
+	// first node in the linked list
+	if node == nil {
+		node := &Node{
+			key:   key,
+			value: value,
+		}
+		h.arr[hash] = node
+		return
+	}
+
+	// find last node in the linked list
+	for node.next != nil {
+		node = node.next
+	}
+
+	node.next = &Node{
+		key:   key,
+		value: value,
+	}
+
+}
+
+func (h *HashTable) Search(key int) (int, error) {
```

Comment:
Search function will return a value if the key exists. Otherwise, it can return `nil`.

Reply:
Unfortunately, we can't return `nil` in place of an integer in Go. The convention is to return another value, which is `error` in this case to determine if the operation succeeds or not.

Reply:
I'm not sure about Go part but normally `search` function should not give an error. @thanks173  How do you think about this issue?

Reply:
Default implementation of Go (using `map`) will just return `nil` or zero value (depending on the value type).
I agree with @trungducc

Reply:
`map` in Go is dynamically typed, keys and values can be of any types. For the sake of simplicity, I'm implementing a Hash Table of Integers, so there's no such thing as `nil`. If I return `-1` then I wouldn't know if it's `nil` or a real `-1` value.

Reply:
It's ok to return a number if the key doesn't exist. There are 2 points here:
- This function should not give an error
- If the key doesn't exist, I prefer to return min value of Int and define it as a public constant.  For example, we can use `search(1) == HashTable.NOT_FOUND` to determine the key exist or not.

Reply:
I think I need help from @thanks173 here. I'm still learning Go so I don't know if there's a convention for declaring enums, or constants in Go.

Reply:
In the default implementation, Go use a two-value assignment to check if a key is present or not, and it's optional.
```
elem, ok = m[not_existed_key] // return nil/0/false for elem, and false for ok
elem = m[not_existed_key] // return nil/0/false for elem
```
For the sake of simplicity, I think returning `0` value is ok. It's the same with initializing an integer without giving it any value.

Reply:
For enums, you could use Go's equivelent `iota`. For the convention, you can read [here](https://yourbasic.org/golang/iota/#iota-basic-example). 

But as the function should return `int`, I don't think it's a good idea to use another `int` to determine the key exist or not.

The zero value (`nil`, `0`, `false`, `""`,...), on the other hand, is actually more useful than you may think. You can read it [here](https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful). In the end (your application), you are dealing with the value stored in the hash table, not the table itself. So using the zero value as a known default is always recommended (i.e, checking hash table for a logged-in user, the value `0` should mean that the user is not logged-in, whether the hash table contains that `user_id` or not).

Your approach is fine btw, it's just complicated.